src/lib/db/api-token-store.ts

```diff
@@ -72,26 +97,64 @@ export class ApiTokenStore implements IApiTokenStore {
 
     async getAll(): Promise<IApiToken[]> {
         const stopTimer = this.timer('getAll');
-        const rows = await this.db<ITokenTable>(TABLE);
+        const rows = await this.makeTokenProjectQuery();
         stopTimer();
-        return rows.map(toToken);
+        return toTokens(rows);
     }
 
     async getAllActive(): Promise<IApiToken[]> {
         const stopTimer = this.timer('getAllActive');
-        const rows = await this.db<ITokenTable>(TABLE)
+        const rows = await this.makeTokenProjectQuery()
             .where('expires_at', 'IS', null)
             .orWhere('expires_at', '>', 'now()');
         stopTimer();
-        return rows.map(toToken);
+        return toTokens(rows);
+    }
+
+    private makeTokenProjectQuery() {
+        return this.db<ITokenRow>(`${TABLE} as tokens`)
+            .leftJoin(
+                `${API_LINK_TABLE} as token_project_link`,
+                'tokens.secret',
+                'token_project_link.secret',
+            )
+            .select(
+                'tokens.secret',
+                'username',
+                'type',
+                'expires_at',
+                'created_at',
+                'seen_at',
+                'environment',
+                'token_project_link.project',
+            );
```

Comment:
I am pretty sure this query can be written to actually return the list of projects as a string array directly from postgres using the "ARRAY_AGG" function. I am not sure if it is more efficient, but would at least avoid the need for the reducer processing the items.

Reply:
Oh cool, I didn't actually know about ARRAY_AGG, that's super cool. I'll have to shuffle the query around but this shouldn't be a very hot path/large dataset so the readability wins out here for me anyway, thanks!